src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,473 @@
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode.Ancestors().Any(n => n.Kind() == SyntaxKind.ForStatement))
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterPosition))
+                {
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterPosition < 0)
+            {
+                lastDelimiterPosition = currentNode.Span.End;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterPosition, currentNode.Kind())));
+            nextCommandHandler();
+        }
+
+        private SyntaxNode GetEnclosingArgumentList(SyntaxNode currentNode, SyntaxToken token, SnapshotPoint? caret, int caretPosition, ISyntaxFactsService syntaxFacts)
+        {
+            if (token.Kind() == SyntaxKind.OpenParenToken && caretPosition == currentNode.SpanStart)
+            {
+                currentNode = currentNode.Parent;
+                if (currentNode == null) return null;
+            }
+            while (currentNode.Kind() != SyntaxKind.ArgumentList
+                    && currentNode.Kind() != SyntaxKind.ArrayRankSpecifier)
+            {
+                if (currentNode.Kind() == SyntaxKind.InterpolatedStringExpression || currentNode.Kind() == SyntaxKind.StringLiteralExpression) return null;
+                if (currentNode == null 
+                    || syntaxFacts.IsStatement(currentNode) 
+                    || currentNode.Kind() == SyntaxKind.VariableDeclaration)  return null;
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                    return null;
+            }
+            // now we're in an argument list, so return the enclosing statement
+            while (!syntaxFacts.IsStatement(currentNode) && !(currentNode.Kind() == SyntaxKind.VariableDeclaration) )
+            {
+                currentNode = currentNode.Parent;
+                if (currentNode == null) return null;
+            }
+
+            return currentNode;
+            
+        }
+
+
+        private bool IsCaretAtEndOfLine(SnapshotPoint? caret, int caretPosition)
+        {
+            return caret.Value.Position == caret.Value.GetContainingLine().End;
+        }
+
+        private SyntaxToken GetToken(SyntaxNode root, int caretPosition, SnapshotPoint? caret)
+        {
+            //previously bailed if caret was null, so this is safe
```

Comment:
💡 Codify statements like this, rather than use comments:

```csharp
Debug.Assert(caret.HasValue, "Should have bailed previously if caret is null");
```